packages/expo-google-sign-in/src/index.js

```diff
@@ -0,0 +1,337 @@
+      this.auth.equals(other.auth) &&
+      this.scopes === other.scopes &&
+      this.hostedDomain === other.hostedDomain &&
+      this.serverAuthCode === other.serverAuthCode
+    );
+  }
+
+  toJSON() {
+    let auth = this.auth;
+    if (this.auth && this.auth.toJSON) {
+      auth = this.auth.toJSON();
+    }
+    return {
+      ...super.toJSON(),
+      auth,
+      scopes: this.scopes,
+      hostedDomain: this.hostedDomain,
+      serverAuthCode: this.serverAuthCode,
+    };
+  }
+}
+
+export type GoogleSignInAuthResult = { type: 'success' | 'cancel', user: ?User };
+
+export class GoogleSignIn {
+  _initialization: Promise;
+  _currentUser: User;
+
+  ERRORS = ERRORS;
+  SCOPES = SCOPES;
+  TYPES = TYPES;
+
+  get currentUser(): User {
+    return this._currentUser;
+  }
+
+  _setCurrentUser(currentUser: ?User): User {
+    this._currentUser = currentUser;
+    return this._currentUser;
+  }
+
+  _validateOptions = (options: ?GoogleSignInOptions): GoogleSignInOptions => {
+    if (!options || !Object.keys(options).length) {
+      throw new Error('GoogleSignIn: invalid configuration');
+    }
+    if (options.offlineAccess && !options.webClientId) {
+      throw new Error('GoogleSignIn: offline use requires server web ClientID');
+    }
+
+    const DEFAULT_SCOPES = [SCOPES.PROFILE, SCOPES.EMAIL];
+    return {
+      ...options,
+      scopes: options.scopes || DEFAULT_SCOPES,
+    };
+  };
+
+  askForPlayServicesAsync = () => this.hasPlayServicesAsync({ showPlayServicesUpdateDialog: true });
+
+  hasPlayServicesAsync = async (options = { showPlayServicesUpdateDialog: false }) => {
+    if (ExpoGoogleSignIn.playServicesAvailableAsync) {
+      if (options && options.showPlayServicesUpdateDialog === undefined) {
+        throw new Error(
+          'ExpoGoogleSignIn: Missing property `showPlayServicesUpdateDialog` in options object for `hasPlayServices`'
+        );
+      }
+      return ExpoGoogleSignIn.playServicesAvailableAsync(options.showPlayServicesUpdateDialog);
+    } else {
+      return true;
+    }
+  };
+
+  initAsync = async (options: ?GoogleSignInOptions): Promise<any> => {
+    this.options = this._validateOptions(options || this.options);
+
+    if (!this.options) {
+      return false;
+    }
+
+    const hasPlayServices = await this.hasPlayServicesAsync();
+    if (!hasPlayServices) {
+      return false;
+    }
+
+    if (this._initialization == null) {
+      try {
+        this._initialization = ExpoGoogleSignIn.initAsync(this.options);
+      } catch (e) {
+        this._initialization = null;
+      }
+    }
+    return this._initialization;
+  };
+
+  _invokeAuthMethod = async (method: string): Promise<?GoogleSignInAuthResult> => {
+    await this.initAsync();
+    const payload = await ExpoGoogleSignIn[method]();
+    let account = payload != null ? new User(payload) : null;
+    return this._setCurrentUser(account);
+  };
+
+  isSignedInAsync = async (): Promise<boolean> => {
+    await this.initAsync();
```

Comment:
Do we have to call `initAsync` before `isSignedInAsync`? Shouldn't `isSignedInAsync` return `false` if not initted?